WIP: RWC POP melody parser

justinsalamon opened the pull request:
See https://github.com/marl/jams-data/issues/1#issuecomment-187388631

justinsalamon committed:
Start work on parser, no functionality yet

justinsalamon commented:
Hey everyone, do you think we could have a folder inside parsers called "data", which contains a folder per parser (when needed) with any data files the parser requires? In my case the RWC Pop melody metadata lives [online in a table](https://staff.aist.go.jp/m.goto/RWC-MDB/rwc-mdb-p.html), which I've converted to a csv file, so I'd like to keep this csv file in the repo.

justinsalamon commented:
On a related note, this table also includes the track duration (resolution in seconds only though). @bmcfee do you think it'd make sense to populate the duration field of the JAMS file based on this table?

I could do it from the audio, but RWC is distributed via CD's (at least it was when I got a copy), which means the exact duration of the audio I have might depend on whatever software I used to rip the CD's in the first place, and I have no way of recovering that information. I also don't have the original CD's at hand. So I was thinking the durations reported online might be my best in this case?

bmcfee commented:
> So I was thinking the durations reported online might be my best in this case?

Yeah, in this case, I think that's the best you can hope for.

RWC is distributed on cds, but I thought they were data (not audio) cds?  I may be mistaken though.

justinsalamon commented:
> RWC is distributed on cds, but I thought they were data (not audio) cds? I may be mistaken though.

I don't remember TBH, you may very well be right. Still, I don't remember by which process I obtained the WAV files that I have... so I'll stick to the times reported online for now.

justinsalamon committed:
Add RWC POP metadata as csv file

Data copied directly from the table that can be found online at:
https://staff.aist.go.jp/m.goto/RWC-MDB/rwc-mdb-p.html

justinsalamon committed:
Implement all functions (see below for details)

Implemented process_folder, create_jams, fill_annotation_metadata and
fill_file_metadata

justinsalamon committed:
Clean up code, update docstrings

justinsalamon committed:
Store remaining RWC metadata in the sandbox

justinsalamon commented:
I think this is good to go, anyone care to CR it before I merge the PR?

justinsalamon commented:
p.s. - I still need to get permission to include the converted jams files in the repo, working on that.

justinsalamon committed:
Fix docstring, remove print

bmcfee commented on parsers/rwcpopmelody_parser.py:
```diff
@@ -0,0 +1,197 @@
+    jam.file_metadata.title = metadata['Title'][n_track]
+
+    d_str = metadata['Length'][n_track]
+    jam.file_metadata.duration = (float(d_str.split(":")[0]) * 60 +
```
this is pretty dirty.  Since we're already using pandas here, we can do something like this instead:

``` python
In [4]: pd.Timedelta('0:2:34').total_seconds()
Out[4]: 154.0
```

(only problem is that you'll have to pad in hours, but i generally prefer that to reinventing second-parsing.)

bmcfee commented on parsers/rwcpopmelody_parser.py:
```diff
@@ -0,0 +1,197 @@
+                                  float(d_str.split(":")[1]))
+
+    # Store remaining RWC metadata in the sandbox
+    sandbox_dict = {
```
seems a little overwrought ; why not just dump all metadata into the sandbox?

Or copy the dict, pop out the ones you don't want, and push the rest?

bmcfee commented on parsers/rwcpopmelody_parser.py:
```diff
@@ -0,0 +1,197 @@
+
+    assert len(track_text) == len(m)
+
+    # Find the melody track: it should START with "melo", or if that's not
```
Is this a standard convention?  Or would it make sense to abstract this out and let a user specify a regular expression to search for instead?  I like `'/melo|voca/'` more than hard-wired logic here.

bmcfee commented on parsers/rwcpopmelody_parser.py:
```diff
@@ -0,0 +1,197 @@
+    elif 'voca' in track_text:
+        index = track_text.index('voca')
+    else:
+        print("ABORTING TRACK: couldn't find melody for: %s" % smf_file)
```
switch away from old-style string formatting please. :)

`"ABORTING .... {}".format(smf_file)`.

bmcfee commented on parsers/rwcpopmelody_parser.py:
```diff
@@ -0,0 +1,197 @@
+        index = track_text.index('voca')
+    else:
+        print("ABORTING TRACK: couldn't find melody for: %s" % smf_file)
+        return 0
```
It's better to raise exceptions than to have specialized return values.

bmcfee commented on parsers/rwcpopmelody_parser.py:
```diff
@@ -0,0 +1,197 @@
+                     format=m.format, tick_relative=m.tick_relative)
+
+    # Write temporary midi file
+    temp_midi = out_file.replace(".jams", ".temp.mid")
```
Can we improve this hack?  Preferably via one of the following:
- Use a properly secure temporary filename
- Skip tempfiles altogether and use a StringIO/BytesIO.  I'm not sure if midi and pretty_midi support that (but they should).

Finally, If any of those asserts below fail, then this will leave temp files on disk.  A better pattern here would be to wrap this all in a `try` block, and put the tempfile cleanup in a `finally` clause.  That way, even if the asserts fail, the cleanup still executes.

bmcfee commented on parsers/rwcpopmelody_parser.py:
```diff
@@ -0,0 +1,197 @@
+    jam.annotations.append(midi_ann)
+
+    # Fill file metadata
+    n_track = int(os.path.basename(smf_file)[4:7]) - 1
```
this seems magical.  What is going on here?  Why are characters 2--5 special?

bmcfee commented on parsers/rwcpopmelody_parser.py:
```diff
@@ -0,0 +1,197 @@
+
+        jams_file = (
+            os.path.join(out_dir,
+                         os.path.basename(smf).replace('.MID', '.jams')))
```
This will do strange things when you have inputs that have `.MID` multiple times.  It would be better to `splitext` and join the desired extension.

bmcfee commented on parsers/rwcpopmelody_parser.py:
```diff
@@ -0,0 +1,197 @@
+        jams_file = (
+            os.path.join(out_dir,
+                         os.path.basename(smf).replace('.MID', '.jams')))
+        jams.util.smkdirs(os.path.split(jams_file)[0])
```
Easier to read as `os.path.dirname(jams_file)`

bmcfee commented:
@justinsalamon C has been R'ed.  Mostly minor style issues, though the tempfile cleanup is a real problem.  Otherwise LGTM.
